File: gfg/POTD/2025-02-10/solution.js
/*
class Node{
    constructor(data){
        this.data = data;
        this.left = null;
        this.right = null;
    }
}
*/

/**
 * @param {Node} root
 * @param {number} K
 * @return {number}
 */

class Solution {
	sumK(root, K) {
		function countPaths(node, currentSum, prefixSumMap) {
			if (!node) return 0

			currentSum += node.data
			let pathCount = prefixSumMap.get(currentSum - K) || 0

			// Update prefix sum map
			prefixSumMap.set(currentSum, (prefixSumMap.get(currentSum) || 0) + 1)

			// Recur for left and right subtrees
			let leftPathCount = countPaths(node.left, currentSum, prefixSumMap)
			let rightPathCount = countPaths(node.right, currentSum, prefixSumMap)

			// Backtrack by removing the current sum entry
			prefixSumMap.set(currentSum, prefixSumMap.get(currentSum) - 1)
			if (prefixSumMap.get(currentSum) === 0) {
				prefixSumMap.delete(currentSum)
			}

			return pathCount + leftPathCount + rightPathCount
		}

		let prefixSumMap = new Map()
		prefixSumMap.set(0, 1)
		return countPaths(root, 0, prefixSumMap)
	}
}

File: gfg/POTD/2025-02-10/solution.py
'''
class Node:
    def __init__(self, val):
        self.right = None
        self.data = val
        self.left = None
'''
class Solution:
    def sumK(self,root,k):
        def countPaths(node, currentSum, prefixSumMap):    
            if node is None:
                return 0

            currentSum += node.data
            pathCount = prefixSumMap.get(currentSum - k, 0)

            # Update prefix sum map
            prefixSumMap[currentSum] = prefixSumMap.get(currentSum, 0) + 1

            # Recur for left and right subtrees
            leftPathCount = countPaths(node.left, currentSum, prefixSumMap)
            rightPathCount = countPaths(node.right, currentSum, prefixSumMap)

            # Backtrack by removing the current sum entry
            prefixSumMap[currentSum] -= 1
            if prefixSumMap[currentSum] == 0:
                del prefixSumMap[currentSum]

            return pathCount + leftPathCount + rightPathCount

        prefixSumMap = {0: 1}
        return countPaths(root, 0, prefixSumMap)